File: LeetCodeTop100CSharp/LeetCodeTop100CSharp/98_ValidateBinarySearchTree.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LeetCodeTop100CSharp {
    class _98_ValidateBinarySearchTree {
        public bool IsValidBST(TreeNode root) {
            var stack = new Stack<TreeNode>();
            var node = root;
            var val = double.NegativeInfinity;
            while (node != null || stack.Count > 0) {
                while (node!= null) {
                    stack.Push(node);
                    node = node.left;
                }
                node = stack.Pop();
                if(val >= node.val) {
                    return false;
                }
                val = node.val;
                node = node.right;
            }

            return true;
        }
    }
}

File: algirithm/98_ValidateBinarySearchTree.go
package algirithm

func isValidBST(root *TreeNode) bool {
	if root == nil {
		return true
	}

	if root.Left != nil {
		if root.Val <= root.Left.Val {
			return false
		}
	}

	if root.Right != nil {
		if root.Val >= root.Right.Val {
			return false
		}
	}

	return isValidBST(root.Left) && isValidBST(root.Right)
}
